File: 0076-minimum-window-substring/0076-minimum-window-substring.js
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {
    const map = {}
    for (const char of t) {
        map[char] = map[char] + 1 || 1
    }
    let remain = t.length
    let l = 0
    let r = 0
    let start = 0
    let len = Infinity
    while (r < s.length) {
        if (map[s[r]] > 0) remain--
        map[s[r]]--
        r++
        while (l < r && remain === 0) {
            if (r - l < len) {
                len = r - l
                start = l
            }
            map[s[l]]++
            if (map[s[l]] > 0) remain++
            l++
        }
    }
    return len === Infinity ? "" : s.substring(start, start+len)
};

File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        mp = {}
        for c in t:
            if c not in mp:
                mp[c] = 0
            mp[c] += 1
        
        lo, hi, start, length, count = 0, 0, 0, inf, len(t)
        
        while hi < len(s):
            if s[hi] in mp:
                if mp[s[hi]] > 0:
                    count -= 1
                mp[s[hi]] -= 1
                
            hi += 1
            
            while lo < hi and count == 0:
                if hi - lo < length:
                    length = hi - lo
                    start = lo
                
                if s[lo] in mp:
                    mp[s[lo]] += 1
                
                    if mp[s[lo]] > 0:
                        count += 1
                
                lo += 1
            
        return s[start:start+length:] if length != inf else ""